Tickets

scottschurr opened the pull request:
This is a big pull request and it affects a lot of stuff.  Thanks in advance to my reviewers for taking on a review of this size.

The pull request is divided into a number of focused commits.  Each commit is targeted at making one significant change.  This is in an effort to make reviewing the changes easier.  During the review we can decide whether the commits should be squashed before going into develop.  That decision will affect how I make the `[FOLD]` commits for fixes.  If the intent is to squash, then I don't need to separate the `[FOLD]` commits by their target commits.  But I can go either way.

One additional decision will be whether we want to keep the top-most commit.  If we want to merge tickets into develop but not have them automatically enable onto the network, then the top-most commit will do that work.  However, if we're good with having tickets enabled onto the network soon (modulo validator amendment voting), then we should remove the top-most commit.

FWIW, this code has been through a number of rebases.  I know that rebases occasionally do unexpected things.  I'm hoping the reviewers will keep their eyes open for things that look funny.  Thanks.

My thanks to everyone who contributed to making tickets work.  🙇

HowardHinnant commented on src/ripple/app/tx/impl/CreateTicket.cpp:
```diff
@@ -23,105 +23,135 @@
-        auto const reserve = view().fees().accountReserve(
-            sle->getFieldU32(sfOwnerCount) + 1);
+        XRPAmount const reserve {view().fees().accountReserve (
+            sleAccountRoot->getFieldU32(sfOwnerCount) + ticketCount)};
```
This is a minor complaint:

I'm not fond of the use of explicit (direct) initialization when implicit (copy) initialization would get the job done.  My rationale for this preference is that authors of types (such as `XRPAmount`) make the less-safe constructors/conversions explicit while making the more-safe ones implicit.  So users should prefer the more-safe ones, reserving the explicit syntax just for those initializations where one is acknowledging that one has to use a less-safe bit of code.  For example:
```
         XRPAmount const reserve = view().fees().accountReserve (
            sleAccountRoot->getFieldU32(sfOwnerCount) + ticketCount);
```
I know that rippled does not currently follow this guideline at all, and indeed, tends towards the opposite.  And this is my first stab at reversing that trend:  by calling out places where we're using the safe syntax, and moving towards the unsafe syntax.  Note that I do not have an issue here with the replacement of `auto` with `XRPAmount`.

Short (7min) presentation on the potential dangers of overuse of the explicit syntax:  https://www.youtube.com/watch?v=hobFOAehwio

scottschurr replied:
Fair, and thanks for pointing this out.  I've been following Nico Josuttis's advice on this, but you make great points.  I'm happy to make the change.

Not to add to your load, but if you could point out all of the places you'd like to see fixed then I can fix them all in a single `[FOLD]` commit, or at least one `[FOLD]` per , um, non-FOLD commit.  Is that a fair request?

HowardHinnant commented on src/ripple/app/tx/impl/CreateTicket.cpp:
```diff
@@ -23,105 +23,135 @@
+        // In C++17: put "ret" into if's init-statement, lose outer braces.
+        NotTEC const ret {preflight1 (ctx)};
+        if (!isTesSuccess (ret))
+            return ret;
```
```
    if (std::uint32_t const count = ctx.tx[sfCount];
            count < minValidCount || count > maxValidCount)
        return {temINVALID_COUNT, conseq};
    if (NotTEC const ret = preflight1(ctx); !isTesSuccess (ret))
        return {ret, conseq};
```

HowardHinnant commented on src/ripple/app/tx/impl/CreateTicket.cpp:
```diff
@@ -23,105 +23,135 @@
+    // too many tickets.
+    auto const optTicketCount {(*sleAccountRoot)[~sfTicketCount]};
+    std::uint32_t const curTicketCount {optTicketCount ? *optTicketCount : 0u};
+    std::uint32_t const addedTickets {ctx.tx[sfCount]};
```
Haven't checked that this will work with copy initialization, but I suspect they will.

HowardHinnant commented on src/ripple/app/tx/impl/CreateTicket.cpp:
```diff
@@ -23,105 +23,135 @@
+    // Each ticket counts against the reserve of the issuing account, but we
     // check the starting balance because we want to allow dipping into the
     // reserve to pay fees.
+    std::uint32_t const ticketCount {ctx_.tx[sfCount]};
```
Copy initialization?

HowardHinnant commented on src/ripple/app/tx/impl/CreateTicket.cpp:
```diff
@@ -23,105 +23,135 @@
+    // root sequence.  Before we got here to doApply(), the transaction
+    // machinery already incremented the account root sequence if that
+    // was appropriate.
+    std::uint32_t const firstTicketSeq {(*sleAccountRoot)[sfSequence]};
```
copy initialization.

HowardHinnant commented on src/ripple/app/tx/impl/CreateTicket.cpp:
```diff
@@ -23,105 +23,135 @@
-            return tesSUCCESS;
+        // Sanity check that the transaction machinery really did already
+        // increment the account root Sequence.
+        std::uint32_t const txSeq {ctx_.tx[sfSequence]};
```
copy initialization.

HowardHinnant commented on src/ripple/app/tx/impl/CreateTicket.cpp:
```diff
@@ -23,105 +23,135 @@
-        AccountID const target_account (ctx_.tx.getAccountID (sfTarget));
+        SLE::pointer sleTicket {
+            std::make_shared<SLE>(ltTICKET,
+                getTicketIndex (account_, firstTicketSeq + i))};
```
copy initialization.

HowardHinnant commented on src/ripple/app/tx/impl/CreateTicket.cpp:
```diff
@@ -23,105 +23,135 @@
+    std::uint32_t const oldTicketCount {[&sleAccountRoot] ()
+        {
+            auto const prevTicketCount {(*sleAccountRoot)[~sfTicketCount]};
+            if (! prevTicketCount)
```
copy initialization and init within if.

HowardHinnant commented on src/ripple/app/tx/impl/CreateTicket.h:
```diff
@@ -31,15 +31,53 @@ class CreateTicket
+    //
+    // The number was chosen arbitrarily and is an effort toward avoiding
+    // ledger-stuffing with Tickets.
+    constexpr static std::uint32_t maxTicketThreshold {250};
```
copy + narrow?

HowardHinnant commented on src/test/app/Ticket_test.cpp:
```diff
@@ -25,369 +25,375 @@ namespace ripple {
-                {expiration ? 1: 2, sfDeletedNode.fieldName, jss::Ticket},
-                {3, sfDeletedNode.fieldName, jss::DirectoryNode}
-            };
+            std::string const txType {tx[sfTransactionType.jsonName].asString()};
```
copy init

HowardHinnant commented on src/test/app/Ticket_test.cpp:
```diff
@@ -25,369 +25,375 @@ namespace ripple {
+            }
         }
-        else
+        std::uint32_t const count {tx[sfCount.jsonName].asUInt()};
```
copy+narrow init

HowardHinnant commented on src/test/app/Ticket_test.cpp:
```diff
@@ -25,369 +25,375 @@ namespace ripple {
+            return;
         }
+        std::uint32_t const txSeq {tx[sfSequence.jsonName].asUInt()};
+        std::string const account {tx[sfAccount.jsonName].asString()};
```
copy/narrow init

HowardHinnant commented on src/test/app/Ticket_test.cpp:
```diff
@@ -25,369 +25,375 @@ namespace ripple {
-
-        // verify the actual metadata against the expected
-        for (auto const& it : expected_nodes)
+        Json::Value const& metadata {env.meta()->getJson (JsonOptions::none)};
```
copy init.

HowardHinnant commented on src/test/app/Ticket_test.cpp:
```diff
@@ -25,369 +25,375 @@ namespace ripple {
+        BEAST_EXPECT (metadata[sfAffectedNodes.fieldName].isArray());
+
+        bool directoryChanged {false};
+        std::uint32_t acctRootFinalSeq {0};
```
copy init

HowardHinnant commented on src/test/app/Ticket_test.cpp:
```diff
@@ -25,369 +25,375 @@ namespace ripple {
+        {
+            if (node.isMember (sfModifiedNode.jsonName))
+            {
+                Json::Value const modified {node[sfModifiedNode.jsonName]};
```
copy init

HowardHinnant commented on src/test/app/Ticket_test.cpp:
```diff
@@ -25,369 +25,375 @@ namespace ripple {
+                        // Verify the account root Sequence did the right thing.
+                        std::uint32_t const prevSeq {
+                            modified[sfPreviousFields.jsonName]
+                                [sfSequence.jsonName].asUInt()};
```
copy+narrow init

HowardHinnant commented on src/test/app/Ticket_test.cpp:
```diff
@@ -25,369 +25,375 @@ namespace ripple {
+                        }
+                    }
+
+                    std::uint32_t const consumedTickets {txSeq == 0u ? 1u : 0u};
```
copy+narrow init

HowardHinnant commented on src/test/app/Ticket_test.cpp:
```diff
@@ -25,369 +25,375 @@ namespace ripple {
+                    // But, since the count did not change, we know it equals
+                    // the final Ticket count.
+                    bool const unreportedPrevTicketCount {
+                        count == 1 && txSeq == 0};
```
copy+narrow init

HowardHinnant commented on src/test/app/Ticket_test.cpp:
```diff
@@ -25,369 +25,375 @@ namespace ripple {
+
+                        std::uint32_t const finalCount {
+                            modified[sfFinalFields.jsonName]
+                                [sfOwnerCount.jsonName].asUInt()};
```
copy+narrow init

HowardHinnant commented on src/test/app/Ticket_test.cpp:
```diff
@@ -25,369 +25,375 @@ namespace ripple {
+                            modified[sfPreviousFields.jsonName].isMember(
+                                sfTicketCount.jsonName) ?
+                            modified[sfPreviousFields.jsonName]
+                                [sfTicketCount.jsonName].asUInt() : 0u};
```
copy+narrow init

HowardHinnant commented on src/test/app/Ticket_test.cpp:
```diff
@@ -25,369 +25,375 @@ namespace ripple {
+            }
+            else if (node.isMember (sfCreatedNode.jsonName))
+            {
+                Json::Value const created {node[sfCreatedNode.jsonName]};
```
copy init

HowardHinnant commented on src/test/app/Ticket_test.cpp:
```diff
@@ -25,369 +25,375 @@ namespace ripple {
+            {
+                Json::Value const deleted {node[sfDeletedNode.jsonName]};
+                std::string const entryType {
+                    deleted[sfLedgerEntryType.jsonName].asString()};
```
copy init

HowardHinnant commented on src/ripple/app/tx/impl/Transactor.cpp:
```diff
@@ -277,19 +306,56 @@ Transactor::checkSeq (PreclaimContext const& ctx)
+        // Delete the Ticket, adjust the account root ticket count, and
+        // reduce the owner count.
+        uint256 const ticketIndex {
+            getTicketIndex (account_, seqOrT.value())};
```
copy+narrow init

HowardHinnant commented on src/ripple/app/tx/impl/Transactor.cpp:
```diff
@@ -277,19 +306,56 @@ Transactor::checkSeq (PreclaimContext const& ctx)
-    auto const sle = view().peek(keylet::account(account_));
-    if (! sle)
-        return;
+    SeqOrTicket const seqOrT {ctx_.tx.getSeqOrTicket()};
```
copy init

HowardHinnant commented on src/ripple/app/tx/impl/Transactor.cpp:
```diff
@@ -277,19 +306,56 @@ Transactor::checkSeq (PreclaimContext const& ctx)
+        }
 
-    std::uint32_t const t_seq = ctx_.tx.getSequence ();
+        std::uint64_t const page {(*sleTicket)[sfOwnerNode]};
```
copy+narrow init

HowardHinnant commented on src/ripple/protocol/impl/STTx.cpp:
```diff
@@ -162,6 +162,20 @@ Blob STTx::getSignature () const
 
+SeqOrTicket STTx::getSeqOrTicket () const
+{
+    std::uint32_t const seq {getSequence ()};
```
copy+narrow init

HowardHinnant commented on src/ripple/app/tx/impl/CreateCheck.cpp:
```diff
@@ -183,12 +183,14 @@ CreateCheck::doApply ()
-    std::uint32_t const seq {ctx_.tx.getSequence()};
+    // Note that we we use the value from the sequence or ticket as the
+    // Check sequence.  For more explanation see comments in SeqOrTicket.h.
+    std::uint32_t const seq {ctx_.tx.getSeqOrTicket().value()};
```
copy+narrow init

HowardHinnant commented on src/ripple/app/tx/impl/CreateTicket.cpp:
```diff
@@ -64,8 +64,16 @@ CreateTicket::preclaim(PreclaimContext const& ctx)
-
-    if (curTicketCount + addedTickets > maxTicketThreshold)
+    std::uint32_t const consumedTickets {
+        ctx.tx.getSeqOrTicket().isTicket() ? 1u : 0u};
```
copy+narrow init

HowardHinnant commented on src/ripple/app/tx/applySteps.h:
```diff
@@ -38,6 +38,138 @@ isTecClaimHardFail(TER ter, ApplyFlags flags)
+        // Both this and nextSeqOrT have sequence numbers.  Requires an
+        // exact match of the expected value.  No gaps in sequence numbers
+        // allowed except when Tickets are created.
+        SeqOrTicket const expectSeq {followingSeq()};
```
copy init

HowardHinnant commented on src/ripple/app/tx/impl/CreateCheck.cpp:
```diff
@@ -183,12 +183,14 @@ CreateCheck::doApply ()
         std::make_shared<SLE>(ltCHECK, getCheckIndex (account_, seq));
 
     sleCheck->setAccountID (sfAccount, account_);
+    AccountID const dstAccountId {ctx_.tx[sfDestination]};
```
copy init

HowardHinnant commented on src/ripple/app/ledger/impl/LocalTxs.cpp:
```diff
@@ -148,14 +147,30 @@ class LocalTxsImp
-            std::shared_ptr<SLE const> sle = view.read(
-                 keylet::account(txn.getAccount()));
-            if (! sle)
+            AccountID const acctID {txn.getAccount()};
```
copy init

HowardHinnant commented on src/ripple/app/ledger/impl/LocalTxs.cpp:
```diff
@@ -148,14 +147,30 @@ class LocalTxsImp
                 return false;
-            return sle->getFieldU32 (sfSequence) > txn.getSeq ();
+
+            std::uint32_t const acctSeq {sleAcct->getFieldU32 (sfSequence)};
```
copy+narrow init

HowardHinnant commented on src/ripple/app/ledger/impl/LocalTxs.cpp:
```diff
@@ -148,14 +147,30 @@ class LocalTxsImp
-            return sle->getFieldU32 (sfSequence) > txn.getSeq ();
+
+            std::uint32_t const acctSeq {sleAcct->getFieldU32 (sfSequence)};
+            SeqOrTicket const seqOrT {txn.getSeqOrTicket()};
```
copy init

HowardHinnant commented on src/ripple/app/misc/impl/TxQ.cpp:
```diff
@@ -353,75 +364,100 @@ TxQ::isFull() const
+    // the very beginning of the queue has a gap (i.e., the first entry
+    // in the queue is greater than the account root's Sequence field).
+    // In that situation we want the caller to use their suggested start.
+    TxQAccount::TxMap::const_iterator prevIter {queued.cend()};
```
copy init

HowardHinnant commented on src/ripple/app/misc/impl/TxQ.cpp:
```diff
@@ -525,10 +565,32 @@ TxQ::tryClearAccountQueue(Application& app, OpenView& view,
         it->second.lastResult = txResult.first;
+
+        // Here's an unexpected hack.
+        //
```
Good comment except for the first two lines. :-)

HowardHinnant commented on src/ripple/app/misc/impl/TxQ.cpp:
```diff
@@ -745,188 +914,237 @@ TxQ::apply(Application& app, OpenView& view,
+            // so we know how much to remove from the account balance
+            // for the trial preclaim.
+            XRPAmount potentialSpend {beast::zero};
+            XRPAmount totalFee {beast::zero};
```
copy init

scottschurr commented:
Thanks for the review @HowardHinnant!  I'll tend to the comments early next week.

cjcobb23 commented on src/ripple/proto/rpc/v1/account_info.proto:
```diff
@@ -56,13 +56,21 @@ message QueueData {
+    uint32 ticket_count = 4;
 
-    XRPDropsAmount max_spend_drops_total = 5;
+    uint32 lowest_sequence = 5;
```
@scottschurr Here is a link describing the types of changes that can and can't be made to preserve backwards compatibility: https://developers.google.com/protocol-buffers/docs/proto3#updating

cjcobb23 commented:
> The protobuf files for gRPC changed significantly in 1.5 b6. We need to rebase on top of that. Once we do that, we need to add support for `CreateTicket` transactions and the `Ticket` ledger object. We also need to add `ticket_sequence` as a field to the `Transaction` message type (I would make this a oneof actually, which can be `ticket_sequence` or `sequence`). This entails adding to the protobuf files, as well as adding C++ code to serialize the new data to protobuf. If you rebase, it will be more apparent what needs to be done, and I can assist in any way necessary.

@scottschurr We also need to update the AccountRoot ledger object in the protobuf files. I missed that one the first time.

scottschurr commented:
Hey reviewers, @cjcobb23 noticed that there will be a bunch of work to do when I rebase to -b6.  I think that may influence how you choose to spend your review time.  Here's my current plan:

1. I'm in the middle of incorporating @HowardHinnant's comments.  I'm going to finish that and get @HowardHinnant's okay before I do the rebase.
2. Then I'll rebase to -b6 before addressing @cjcobb23's comments.
3. Then I will address @cjcobb23's comments.

I'm going to suggest that the other reviewers might want to hold off (if they haven't started yet) until after the -b6 rebase.  Thanks, and sorry for the disruption.

scottschurr commented:
@HowardHinnant I just pushed a bunch of `[FOLD]` commits that were intended to address all of your comments.  Unfortunately I just noticed that I missed one comment at the top of CreateTicket.h (line 34).  It would be great if you could review the `[FOLD]` commits to make sure I addressed your concerns (except for the one).  If you're good with the changes then I can squash the `[FOLD]` commits into place (and fix the one that I missed).  I'd like to do that squash before I rebase this pull request to -b6.  Thanks for the help.

HowardHinnant commented:
Still 👍 for me.

scottschurr commented on src/ripple/protocol/SeqOrTicket.h:
```diff
@@ -0,0 +1,157 @@
+     may only be used once there will never be any duplicates within this
+     account.
+*/
+class SeqOrTicket
```
Yup, names are hard.  I struggled to get this one.  It was not off the cuff.  But I also acknowledge that it's not great.  I'm open to ideas.

seelabs replied:
I probably like `seqProxy` the best. I do think `SeqOrTicket` needs to be replaced with something tho.

scottschurr commented on src/ripple/protocol/SeqOrTicket.h:
```diff
@@ -0,0 +1,157 @@
+    Type type_;
+
+public:
+    constexpr explicit SeqOrTicket (Type t, std::uint32_t v)
```
_Almost_ every situation that wants a `Sequence` can also take a `Ticket`.  There are a few exceptions.  But I started out by trying to handle them as separate concepts and it was not working.

Every valid transaction comes in with exactly one of `Sequence` or `Ticket`. You could put three interfaces on a transaction that return `SeqOrTicket`, `Sequence`, and `Ticket`.  But there's no _static_ safety there.  The best you can do is throw an exception if the wrong type is asked for.  And _usually_ you don't have to tell them apart.  So providing a single interface that can always be satisfied (a `SeqOrTicket`) seemed like the right answer.  And I think it plays out well in the code.

I could try having a conversion operator hierarchy like we have with `TEC` codes.  Then we'd have `Sequence`, `Ticket`, and `SeqOrTicket`.  But we would have very few uses of `Sequence` or `Ticket`.  And for the most part you'd get to `Sequence` or `Ticket` with a runtime check followed by an explicit cast.  I suspect it would not improve the situation much.  But I could be wrong.

seelabs replied:
I would not have a conversion hierarchy - I agree, that's overkill. I'll take a closer look at making sequence and ticket distinct types, you might be right that it's not worth it.

scottschurr commented on src/ripple/app/tx/applySteps.h:
```diff
@@ -38,6 +39,134 @@ isTecClaimHardFail(TER ter, ApplyFlags flags)
+    /// Move assignment operator
+    TxConsequences& operator=(TxConsequences&&) = default;
+
+    /// Fee
```
This is a commenting style that was already in the file.  I just retained the style.  I think it's a doxygen thing, but I'm not sure.

scottschurr commented on src/ripple/app/tx/impl/CashCheck.h:
```diff
@@ -34,7 +34,7 @@ class CashCheck
 
     static
-    NotTEC
+    std::pair<NotTEC, TxConsequences>
```
FWIW, `TxConsequences` is not just about blocking.  It's also about spending.  And now it also carries information about `Sequence` jumps.

What you're proposing could be done.  But I think the cost would be an extra `Transactor` entry point that returned `TxConsequences`.  That entry point would _always_ be called just after `preflight`.

Personally I like forcing the author of each `preflight` method to think about what their consequences are.  If `consequences()` were a separate method with a defaulted behavior, then it becomes easy for a Transactor author to not think about it and get the default behavior, which may not be appropriate.  But that's just my approach, and I acknowledge that there are other approaches.

seelabs replied:
I might look at coding something up, to see how hard it would be (if I get some time, I'm behind on some other stuff right now). I do know it's more than just about blocking, but the added logic in `preflight` at least looked like it would be better centralized in `applySteps` rather than spreading it around. I'll try to take a closer look, it think it would be a win if we could do that.

seelabs replied:
@scottschurr 
Of our 20 transactors:
* 12 of them create `TxConsequences` as `TxConsequences{ctx.tx}`
* 3 of them create `TxConsequences` as `TxConsequences{ctx.tx,TxConsequences::blocker};`
* 5 of them create them do something else.

After the consequences are created, all the transactors just return the object.

The problem I have with the current solution is it modifies every `preclaim` and changes the return type of `preclaim`, it also changes `preclaim` the same way in every transactor, and in 15 of 20 cases creates the `TxConsequences` object in the same way. Here's two ideas that try to fix this:

1) Every transaction has to define an `static constexpr ConsequencesFactory` that has one of three values: `Normal`, `Blocking`, and `Custom`. When it's `Custom`, then the transactor has to define a `MakeTxConsequences` static function. With this solution, the existing `preflight` functions do not require any modification. If a transactor doesn't define this variable, a compile error happens. This forces transactor authors to think about `TxConsequences`.

2) As you suggest, we can define an extra `Transactor` entry point. We could have a default one in `Transactor` that handles non-blocking consequences, and a transactor would define their own if they wanted to override this. (we could also not provide a default, if we wanted a compile error).

Branch with patch 1 (enum): https://github.com/seelabs/rippled/commits/consequences-enum
Branch with patch 2 (entry point with default): https://github.com/seelabs/rippled/tree/consequences-static-funs
Of these, I would rank them as:
1) Enum per transactor
2) `MakeTxConsequences` entry point _without_ a default in the base class (no patch provided, but it's easy to see how this would work from patch 2)
3) `MakeTxConsequences` entry point with a default.
4) Modify all the `preclaim` functions

Please note that I just coded these patches up to give an idea what the solution would look like. I'd want to review them carefully myself if we do decide to cherry pick one of them.

scottschurr replied:
The `enum` approach is clever and not overly subtle.  Nice use of `if constexpr` which gives a bunch of compile-time validation.  Sure, I'd be fine switching to that approach.  Thanks for coding it up. :+1:

scottschurr replied:
I have the enum approach integrated.  And it _almost_ works.

It looks like an old Visual C++ compiler is not good enough at `if constexpr` to support that approach.  The three Travis builds that use MSVC 19.16.27035.0 are failing with messages like this (see https://travis-ci.com/github/scottschurr/rippled/jobs/305215758):
```
C:/Users/travis/build/scottschurr/rippled/src/ripple/app/tx/impl/applySteps.cpp(56): error C2039: 'makeTxConsequences': is not a member of 'ripple::DeleteAccount'
..\src\ripple/app/tx/impl/DeleteAccount.h(29): note: see declaration of 'ripple::DeleteAccount'
C:/Users/travis/build/scottschurr/rippled/src/ripple/app/tx/impl/applySteps.cpp(73): note: see reference to function template instantiation 'std::pair<ripple::NotTEC,ripple::TxConsequences> ripple::invoke_preflight_helper<ripple::DeleteAccount>(const ripple::PreflightContext &)' being compiled
C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\VC\Tools\MSVC\14.16.27023\include\chrono(861): note: see reference to class template instantiation 'std::chrono::duration<__int64,std::ratio<1,1>>' being compiled
C:/Users/travis/build/scottschurr/rippled/src/ripple/app/tx/impl/applySteps.cpp(56): error C3861: 'makeTxConsequences': identifier not found
C:/Users/travis/build/scottschurr/rippled/src/ripple/app/tx/impl/applySteps.cpp(56): error C2039: 'makeTxConsequences': is not a member of 'ripple::CancelCheck'
..\src\ripple/app/tx/impl/CancelCheck.h(27): note: see declaration of 'ripple::CancelCheck'
C:/Users/travis/build/scottschurr/rippled/src/ripple/app/tx/impl/applySteps.cpp(77): note: see reference to function template instantiation 'std::pair<ripple::NotTEC,ripple::TxConsequences> ripple::invoke_preflight_helper<ripple::CancelCheck>(const ripple::PreflightContext &)' being compiled
```
All of the other builds, including the build using VC2019, are working.  So Microsoft knows they have a bug and they've fixed it.  But, at least for now, it looks like that technique is not gonna be useable by us.

I may be able to get the same effect (in a way that all of our compilers support) by specializing templates.  That's what I'm gonna try next.

Bummer.

scottschurr commented on src/ripple/app/tx/impl/CreateTicket.cpp:
```diff
@@ -23,105 +23,137 @@
+    // too many tickets.
+    std::uint32_t const curTicketCount =
+        {(*sleAccountRoot)[~sfTicketCount].value_or (0u)};
+    std::uint32_t const addedTickets = {ctx.tx[sfCount]};
```
I think we should include @HowardHinnant in this discussion.

HowardHinnant replied:
I recommended this style.  But it _is_ just a guideline.  I have no intention of trying to enforce it.  However I'm happy to share my rationale.

Yes, `{}` for the narrowing catch.

```c++
uint32_t x = {y};
```

The `=` part is to catch at compile-time an accidental explicit conversion from `y` to `x`.  For example if `y` were to accidentally have the type `std::chrono::day`, then the above would not compile.  However this would compile:

```c++
uint32_t x{y};
```

The author of the type `day` recognized that implicit conversions from `day` to `unsigned` would be dangerous.  But nevertheless the conversion is still useful, so it was made explicit.  By having the client use implicit syntax, the client is saying:  I only want this to happen if it is a safe conversion (or no conversion at all).

With the syntax as coded, we don't have to check if `(*sleAccountRoot)[~sfTicketCount].value_or (0u)` returns a class type with an `explicit operator unsigned()`.  We have more confidence that it is returning an integral type.  And with the `{}`, an unsigned integral type that won't overflow when assigned to `uint32_t`.

Now as coded, closer inspection of `(*sleAccountRoot)[~sfTicketCount].value_or (0u)` makes it likely that this is returning `optional<unsigned>`.  But the point of the guideline is that as one gets used to the `integral x = {y}` syntax, one doesn't even have to inspect `y` to know it is a safe, non-narrowing initialization.

seelabs replied:
Thanks @HowardHinnant I understand the rational. I'm still against doing this in isolation like this. I'm not against making this a recommended way of doing initialization from now on - tho we should have a discussion among the team first.

Also, I apologize if my comment about this being an "experiment" came across poorly.

HowardHinnant replied:
No it didn't come across poorly.  No apologies needed.  I was just trying to give background:  As in I didn't come up with this last week.  Been thinking about this and experimenting for quite a while.

About isolation:  Every marathon begins with one step. :-)  I think dedicating time to a sweeping change in style across all of rippled for any reason is not the best use of our limited time.  I recommend applying best practices to the code you're working on today.  This enables better practices to propagate more widely, and poor practices to die out, with little investment in time.

seelabs replied:
@HowardHinnant @scottschurr I've thought about this, and I'm still against it. The benefit of this `int var = {foo()}` idiom over `int var{foo()}` is it protects against accidentally invoking an `explicit operator int()`. That is realistically never going to happen in this case.

If we are going to introduce this idiom, it would be good if the first place it was introduced had a realistic chance of catching an error - in some generic code say. If we find that place and think we want to change style from now forward, then we talk with the team and make the style change. As is, I still think the cost of making the code different from the rest of rippled outweighs the benefit here.

As usual, there's no "right" answer to this. I suspect I'm weighing code consistency more heavily than you two, and you two are trying to introduce a safer initialization mechanism. While I'm still a "vote no" for the code here, I also understand if it's kept as-is.

HowardHinnant replied:
This issue isn't going to impact my review one way or the other (sorry I'm running behind on that).  @scottschurr, sorry if I had you do a lot of work for nothing.  This looks like your call as to which way to resolve this, and however you do, I'm not loosing any sleep over it.

scottschurr commented on src/ripple/app/misc/impl/TxQ.cpp:
```diff
@@ -356,75 +364,100 @@ TxQ::isFull() const
+{
+    // Find the transaction in the queue that immediately precedes or
+    // equals start.
+    assert (start.isSeq());
```
This is one of the few places where we can only accept a `SeqOrTicket` that is a `Sequence`.  I looked around, and it turns out this function is only called from one place: `TxQ::nextQueuableSeq()`.  _That_ function receives a `uint32_t` and constructs a `SeqOrTicket` out of it, specifying that it is a `Sequence`.

The simplest way out of the mess is to get rid of `TxQ::preceedsNextOpenSeq()` and drop its contents wholesale into `TxQ::nextQueuableSeq()` as a lambda.  Yeah, it makes `TxQ::nextQueuableSeq()` into a bigger method.  But it gets rid of the passed `lock` and gets rid of the `assert`.

The resulting `TxQ::nextQueuableSeq()` might look like this:

```
SeqOrTicket
TxQ::nextQueuableSeq (
    ReadView const& view,
    AccountID const& account, std::uint32_t proposedSeq) const
{
    SeqOrTicket const proposed {SeqOrTicket::seq, proposedSeq};

    std::lock_guard<std::mutex> lock(mutex_);

    // If the account is not in the queue, the proposed start is good enough.
    auto accountIter = byAccount_.find(account);
    if (accountIter == byAccount_.end() ||
        accountIter->second.transactions.empty())
        return proposed;

    // Find the transaction in the queue that precedes the next opening
    // in the queue that follows proposed.
    TxQAccount::TxMap const& txns {accountIter->second.transactions};
    TxQAccount::TxMap::const_iterator const precedes = [proposed, &txns] ()
    {
        // Find the transaction in the queue that immediately precedes or
        // equals proposed.
        TxQAccount::TxMap::const_iterator txIter =
            [proposed, &txns] ()
            {
                auto sameOrNextIter = txns.lower_bound (proposed);
                if (sameOrNextIter != txns.begin())
                    --sameOrNextIter;
                return sameOrNextIter;
            }();

        // Initializing prevIter with txns.cend() covers the case where
        // the very beginning of the queue has a gap (i.e., the first entry
        // in the queue is greater than the account root's Sequence field).
        // In that situation we want the caller to use their proposed.
        TxQAccount::TxMap::const_iterator prevIter = txns.cend();
        for (SeqOrTicket attempt = proposed; txIter != txns.cend(); ++txIter)
        {
            // We're looking for the first hole in the queue that follows
            // proposed.  If attempt is less than txIter's seqOrT, then we've
            // found a hole.
            if (attempt < txIter->first)
                break;

            // Try the next SeqOrTicket that could validly follow txIter.
            // Remember that sequence numbers jump when a transaction
            // creates tickets, so we can't simply increment attempt.
            attempt = txIter->second.consequences().followingSeq();
            prevIter = txIter;
        }
        return prevIter;
    }();

    return precedes == txns.cend() ?
        proposed : precedes->second.consequences().followingSeq();
```

Thoughts?

seelabs replied:
@scottschurr I prefer making a lambda, as you suggest. I'd rather make the function larger than have a free function with a misleading interface.

cjcobb23 replied:
@scottschurr I prefer the lambda as well. I think the resulting size of the function is fine.

seelabs commented on src/ripple/app/tx/impl/PayChan.cpp:
```diff
@@ -170,29 +170,30 @@ closeChannel (
+std::pair<NotTEC, TxConsequences>
 PayChanCreate::preflight (PreflightContext const& ctx)
 {
+    TxConsequences const conseq {ctx.tx};
```
Should this have a potential spend in the consequences?

seelabs commented on src/ripple/app/tx/impl/PayChan.cpp:
```diff
@@ -293,23 +297,24 @@ PayChanCreate::doApply()
+std::pair<NotTEC, TxConsequences>
 PayChanFund::preflight (PreflightContext const& ctx)
 {
+    TxConsequences const conseq {ctx.tx};
```
Should this have a potential spend in the consequences?

scottschurr commented on src/ripple/rpc/handlers/AccountInfo.cpp:
```diff
@@ -120,48 +120,70 @@ Json::Value doAccountInfo (RPC::JsonContext& context)
+                        seqCount += 1;
+                        if (!lowestSeq)
+                            lowestSeq = tx.seqOrT.value();
+                        highestSeq = tx.seqOrT.value();
```
@cjcobb23, you're right.  It does rely on the order that transactions are retrieved by the `TxQ`.  There are a couple of ways to approach that:

1. Add a comment here.
2. Add a comment on `TxQ::getAccountTxs` describing the guaranteed return order.
3. Add a couple of asserts here (and a couple of tracking variables to help with the asserts).
4. Preemptively sort here in case the return order changes in the future.

I'm fine with any of those.  Preemptively sorting is maybe a little expensive, but the entry point is probably not called that often so I don't think it's prohibitive.  Do you have a preferred approach?  Thanks.

cjcobb23 replied:
@scottschurr I agree preemptive sorting might be a little expensive. I think adding a comment here and a couple of asserts would be sufficient.

scottschurr commented on src/ripple/app/misc/impl/TxQ.cpp:
```diff
@@ -767,193 +912,237 @@ TxQ::apply(Application& app, OpenView& view,
+        boost::optional<OpenView> openView;
+    };
+
+    boost::optional<MultiTxn> multiTxn;
```
Nice simplification!  Thanks.

cjcobb23 commented:
> @cjcobb23 and @seelabs, thanks for the great review comments. I have a branch that has integrated most of your comments. Assuming it passes CI I'll be pushing it here tomorrow. I think I've addressed all of your comments except for five. I'd like to discuss those remaining five.
> 
> There are two small changes that I'd like to discuss before I address them:
> 
> 1. @cjcobb23 left a comment about how AccountInfo.cpp makes presumptions regarding the order of returned transactions (https://github.com/ripple/rippled/pull/3271/files#r387180959).  Is adding a comment to that file sufficient?  I could go further and add a couple of asserts.  Or I could go even further and sort the returned values before processing.  Your preference?  Thanks.
> 2. @seelabs suggested that TxFormats.cpp have common fields for regular transactions and different common fields for pseudo transactions (https://github.com/ripple/rippled/pull/3271/files#r394020028).  I could do that, but I'm disinclined.  I'm utterly terrified by the way that TxFormats.cpp and ledger_objects.proto are conceptually joined at the hip but completely disconnected in the code.  I think the common fields in TxFormats.cpp aggravate that condition.  Having distinct common fields makes that yet worse.  But I'm open for discussion.
> 
> There are three big requested changes that are worth discussing as well:
> 
> 1. @seelabs, the `enum` solution for `TxConsequences` looks great.  What is the way you'd like to see it introduced into the commit stack?  If the expectation is that we'll squash the commit stack before merging, then it doesn't make any difference.  The benefit of not squashing is that a bisect has more resolution.  But I don't know whether that's worth while.  So, should I just append the `enum` change to the commit stack, or should I insert it into the correct place (as a `FOLD` to the "Return TxConsequences from preflight" commit).
> 2. @seelabs suggests there is probably a better name for `SeqOrTicket` (https://github.com/ripple/rippled/pull/3271/files#r393986670).  Two issues: 1) deciding on a better name and 2) if we change the name how to apply the change (insert in commit stack or append to the stack.).  Here are some potential names:
>    
>    * `SeqOrTicket`
>    * `SeqProxy`
>    * `TxNum`
>    * `TxOrdinal`
>    
>    Of those I still prefer `SeqOrTicket`.  To me, it says what it is.  If we grasp for an underlying concept, then maybe `TxOrdinal`.  But that name doesn't tie me back to the `Sequence` name we've been using for years.  So I think there's a familiarity curve to struggle with.  But I'm fine with changing to whatever name the group settles on.
> 3. Both @cjcobb23 and @seelabs think it's time to refactor `TxQ::apply()`.  I can't disagree; it's huge.  But I'm not convinced I'm the right guy for the job.  Still, I'm going to make a run at it.  We can decide if we want to keep it once it's done.  I'm assuming that commit should just be appended to the other commits.

@scottschurr in regards to 1 (the order of transactions), I think a comment and a couple of asserts would be fine. As to 3 (refactoring `TxQ::apply()`), I would just make a run at it and keep that change in its own commit, just as you said. You don't have to go overboard on the refactor and do anything fancy; I would just try to break it into a couple of functions. Thanks for taking a stab at it.

seelabs commented on src/ripple/app/tx/applySteps.h:
```diff
@@ -38,6 +39,134 @@ isTecClaimHardFail(TER ter, ApplyFlags flags)
+private:
+    /// Describes how the transaction affects subsequent
+    /// transactions
+    Category category_;
```
Making the member variable `Category category_` instead of `bool isBlocker_` feels like an extra level of indirection. It's easy to see what the `bool` variable is all about. The `Category` variable requires looking at the `Category` enum. But it's no problem if you prefer the `Category` version. It's a minor detail one way or the other.

(I agree the enum should be kept for the constructor argument)

seelabs commented on src/ripple/protocol/impl/TxFormats.cpp:
```diff
@@ -54,6 +54,7 @@ TxFormats::TxFormats ()
             { sfSetFlag,             soeOPTIONAL },
             { sfClearFlag,           soeOPTIONAL },
             { sfTickSize,            soeOPTIONAL },
+            { sfTicketSequence,      soeOPTIONAL },
```
I saw your message in this comment: https://github.com/ripple/rippled/pull/3271#pullrequestreview-378709461 I'm fine not making this change. I'm marking this resolved.

seelabs commented:
@scottschurr Re: https://github.com/ripple/rippled/pull/3271#pullrequestreview-378709461

1) I'd squash this all down. I think the individual commits are useful for reviews, but I don't think the feature works without all the commits, so individual commits wouldn't really help when bisecting.

2) I would rename this. I'm fine with any of the suggestions. I probably prefer `SeqProxy` myself, but the others are fine with me too. My objection to names like `xOrY` is it's not a single concept, I really want a name that's a single concept, whatever that is.

3) It probably makes sense to make refactoring `apply` it's own PR. I don't think you need to tackle that in this PR.

ximinez commented:
I like `SeqProxy`, too. It conveys that the value is still tied to the transaction's sequence number (one could think of a Ticket number as a deferred sequence number), and indicates that it's not a direct access - there's something more to it: the type.

cjcobb23 commented:
@scottschurr I see that you rebased and made some changes to the gRPC protobuf files as well as GRPCHelpers.cpp. Are you finished making the necessary changes for gRPC or are you still working on them?

scottschurr commented:
@cjcobb23, I _hope_ I've made all of the necessary changes to the protobuf files and gRPCHelpers.  Please let me know if/what I missed.  Thanks.

scottschurr commented:
The most recent push includes a number of fixes and a rebase to 1.5.0.  At this point I believe I've addressed all comments to date with one exception: I have not refactored `TxQ::apply()`.  I'd prefer to make that change in a separate pull request if that's okay with everyone.

Thanks very much for all the help so far.  In particular, a shout out to both @seelabs and @cjcobb23 for the great review comments and guidance on the right way to fix things.

seelabs commented:
@scottschurr I agree that refactoring `TxQ::apply` belongs in it's own PR.

seelabs commented on src/ripple/app/tx/impl/applySteps.cpp:
```diff
@@ -39,56 +38,122 @@
+{
+    return {T::preflight(ctx), T::makeTxConsequences(ctx)};
+};
+
```
I see that this was changed to using `enable_if` instead of `if constexpr` (from the origional patch). I think I prefer using `if constexpr` in a single function over three overloads and `enable_if`, just for code clarity. 

I _suspect_ that the reason is I put an `assert` instead of an `static_assert` for if a new `ConsequencesFactory` is added, and the `enable_if` implementation will give a compile error in this case. If that's true, I think I can modify the `if constexpr` solution to use `static_assert` by calling `static_assert(sizeof(T) == -1, ""). I.e. something like:
```
template <class T>
std::pair<NotTEC, TxConsequences>
invoke_preflight_helper(PreflightContext const& ctx)
{
    TxConsequences const conseq = [&]() {
        if constexpr (T::ConsequencesFactory == Transactor::Normal)
        {
            return TxConsequences{ctx.tx};
        }
        else if constexpr (T::ConsequencesFactory == Transactor::Blocker)
        {
            return TxConsequences{ctx.tx, TxConsequences::blocker};
        }
        else if constexpr (T::ConsequencesFactory == Transactor::Custom)
        {
            return T::MakeTxConsequences(ctx);
        }
        else
        {
            static_assert(sizeof(T) == -1, "Unknown consequences factory type");
        }
    }();
    return {T::preflight(ctx), conseq};
}
```
What do you think?

scottschurr replied:
Actually I really like (and prefer) the `if constexpr` approach you coded up.  But when I ran that approach through CI, the Visual Studio complier choked on it.  GCC and Clang were okay.  And the latest Visual Studio (2019?) was okay with it.  But the older current Visual Studio couldn't handle it.

So I patched the code up using the `enable_if` approach.  It's harder for humans to read, but all of the compilers could chew through it.  It's a shame.  The `if constexpr` approach is way easier to read.

seelabs replied:
I don't think we can justify upping visual studio support to 2019 just for this change, but yeah, I agree it's a shame we have work around this limitation.

scottschurr commented:
After consulting with @ximinez, I squashed all of the previous fixes into their appropriate commits.  We're hoping that will make the coder review easier.

As it stands, all of the individual commits build correctly, pass unit tests, and sync to the network using my laptop (macOS).

seelabs commented:
@scottschurr I don't object to multiple commits in a single PR, but I don't want the commits to reflect intermediate development milestones. For example, there's a commit to `Return TxConsequences from preflight:` and a later commit to `Take TxConsequences out of all the preflight functions`. There's also a commit that introduces `SeqToTicket` and another commit that renames this later. Ideally, when this is merged we shouldn't see the rename or the intermediate changes to preflight.

I also need to think about how these intermediate milestone interact with git bisect. Ideally, git bisect should point to the commit that introduced a bug, and I'm not sure these commits do that (tho I admit I haven't looked closely). For example, if we have one commit that introduces a feature with a bug, but doesn't turn it on, and a separate commit that turns the feature on. Git bisect will always point to the commit that turned the feature on as the commit with the bug, not the code that introduced the feature.

I do think we need to resolve the first set of issue (intermediate development artifacts). I have to think about the second - there are pros and cons, as usual.

scottschurr commented:
@seelabs, I don't intend to insist that the commits as they stand all be merged separately.  If we decide to squash them all I'm fine with that decision.  But I wouldn't want to do that before the review by @ximinez.

In terms of the current separate commits, it was a tradeoff regarding safety of squashing the fixes in without unexpected side effects.  I couple of the fixes affected a significant number of files (e.g., renaming `SeqOrTicket` to SeqProxy`).  There was pretty high likelihood of rebase issues if I put those where they belong.  So I left them at the end of the commit stack.  As it is I was fortunate enough that none of the squashing I did caused any rebase conflicts.

scottschurr commented:
Rebased to 1.6.0-b1.

scottschurr commented on src/ripple/app/tx/impl/CreateTicket.h:
```diff
@@ -31,15 +31,53 @@ class CreateTicket
+    // So, for the moment, a single transaction creating 250 Tickets takes
+    // about the same compute time as a single compute-intensive payment.
+    //
+    // October 2018.
```
Somehow leaving my name behind didn't feel right, so I removed the pronouns from the comment.

scottschurr commented on src/ripple/proto/org/xrpl/rpc/v1/transaction.proto:
```diff
@@ -55,8 +55,9 @@ message Transaction
         SignerListSet signer_list_set = 28;
 
-        TrustSet trust_set = 29;
+        TicketCreate ticket_create = 30;
 
+        TrustSet trust_set = 29;
```
Because `TicketCreate` was added after `TrustSet`, but the entries are still listed alphabetically.

scottschurr commented on src/test/app/Ticket_test.cpp:
```diff
@@ -25,369 +25,376 @@ namespace ripple {
+
+                        if (txSeq == 0)
+                        {
+                            // Transaction used a TicketSequence.
```
I don't think I understand this one.  We already know that `txSeq == 0`.  I'm not sure what I gain by checking that `prevSeq != 0`?

ximinez replied:
I think I meant to compare it to the ticket sequence. `BEAST_EXPECT (prevSeq != ticketSeq);` but I don't know if that's already defined or if you'll have to define it. Basically my idea was to demonstrate that a ticket value and sequence values are mutually exclusive.

scottschurr replied:
That would be a useful check, but I think it's more applicable in the `checkTicketConsumeMeta()` function.  Even then it's hard to verify mutual exclusion between all tickets consumed and all consumed sequences.

What I did, which is not nearly as good as what you're asking for, is in `checkTicketConsumeMeta()`, verify that every consumed ticket is less than the account root `Sequence`.

scottschurr commented on src/test/app/Ticket_test.cpp:
```diff
@@ -25,369 +25,376 @@ namespace ripple {
+        {
+            if (node.isMember (sfModifiedNode.jsonName))
+            {
+                Json::Value const modified = node[sfModifiedNode.jsonName];
```
Great idea!  That helps considerably with the readability of that portion of the test.

scottschurr commented on src/test/app/Ticket_test.cpp:
```diff
@@ -25,369 +25,376 @@ namespace ripple {
+                        // the previous TicketCount should not be reported.
+                        BEAST_EXPECT (!
+                            modified[sfPreviousFields.jsonName].isMember(
+                                sfTicketCount.jsonName));
```
Unfortunately it's not that simple.  There are additional conditions that influence whether final fields contains a `TicketCount`.

ximinez replied:
True, but in the context of this test, it's always there. I think a check makes sense in case that ever changes.

scottschurr commented on src/test/app/Ticket_test.cpp:
```diff
@@ -25,369 +25,376 @@ namespace ripple {
+                            modified[sfPreviousFields.jsonName]
+                                [sfTicketCount.jsonName].asUInt() : 0u};
+
+                        if (startCount == 0u)
+                            BEAST_EXPECT (!
+                                modified[sfPreviousFields.jsonName].isMember(
+                                    sfTicketCount.jsonName));
```
I think `||` is the wrong operator to apply here.  Instead we need an exclusive OR.  So I've changed the test like this:
```
                        BEAST_EXPECT (
                            (startCount == 0u) ^
                            previousFields.isMember(sfTicketCount.jsonName));
```

ximinez replied:
Nice!

scottschurr commented on src/test/jtx/impl/ticket.cpp:
```diff
@@ -26,32 +26,13 @@ namespace jtx {
-    jv[jss::TransactionType] = jss::TicketCancel;
-    jv[jss::Account] = account.human();
-    jv["TicketID"] = ticketId;
+    jv[sfCount.jsonName] = count;
```
I'm getting rid of `sfCount` at your suggestion, so the point is moot.  However I do have an opinion about putting `SField` names in `jss`.  I'm mildly against it.  I prefer that there be just one source for the textual name.  If it's an `SField` it has its own name.  The other JSON fields don't have a better source, so `jss` is a good place to get them.  But that's just my take on it...

ximinez replied:
Yeah, that makes a lot of sense.

scottschurr commented:
Appended a `[FOLD]` commit for fixes covering @ximinez's comments on the first commit.

scottschurr commented on src/ripple/protocol/SeqOrTicket.h:
```diff
@@ -0,0 +1,157 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2018 Ripple Labs Inc.
```
2018 is when I wrote the code initially.  I vaguely remember hearing that leaving the older date had some kind of advantage when it comes to proving no prior art.  But I'm no lawyer...

scottschurr commented on src/ripple/protocol/impl/TER.cpp:
```diff
@@ -96,6 +96,7 @@ transResults()
         { tefBAD_AUTH_MASTER,        { "tefBAD_AUTH_MASTER",       "Auth for unclaimed account needs correct master key."                          } },
         { tefINVARIANT_FAILED,       { "tefINVARIANT_FAILED",      "Fee claim violated invariants for the transaction."                            } },
         { tefTOO_BIG,                { "tefTOO_BIG",               "Transaction affects too many items."                                           } },
+        { tefNO_TICKET,              { "tefNO_TICKET",             "Ticket was never created or has already been used."                            } },
```
How about "Ticket is not in ledger."?

ximinez replied:
Yeah, that's great.

scottschurr commented on src/test/rpc/Submit_test.cpp:
```diff
@@ -185,8 +185,9 @@ class Submit_test : public beast::unit_test::suite
-            BEAST_EXPECT(client.reply.engine_result_code() == -198);
+            BEAST_EXPECT(
+                client.reply.engine_result().result() == "tefPAST_SEQ");
+            BEAST_EXPECT(client.reply.engine_result_code() == -190);
```
I was not aware of that, but you are correct.  I believe that's because `tefALREADY` is no longer accessible by a normal transaction.  The sequence check happens first in `invoke_preclaim` which returns `tefPAST_SEQUENCE` before calling `checkPriorTxAndLastLedger()`.  Any transaction that meets the requirements for `tefALREADY` also meets the requirements for `tefPAST_SEQUENCE`.  And the sequence check is cheaper to perform, so it makes sense to do that first.

It looks like we could gin up a unit test to exercise `tefALREADY` using pseudo transactions.  But that would be in a different test file.

In effect I don't think we've actually lost any test coverage here, but we have identified some (currently) inaccessible code.

ximinez replied:
:thinking: Yeah, that makes sense. I'm not too worried about it. (context: `tefALREADY`)

scottschurr commented on src/ripple/app/misc/impl/TxQ.cpp:
```diff
@@ -196,24 +196,35 @@ TxQ::FeeMetrics::scaleFeeLevel(Snapshot const& snapshot,
 
+    // We expect that size_t == std::uint64_t but, just in case, guarantee
+    // we lose no bits.
+    std::uint64_t x {xIn};
```
I prefer not to (intentionally) do numeric type conversions in call parameters.  That is one of several places where C++ silently does narrowing conversions.  The compiler may emit warnings, but is not required to.

Take a look at this program which silently converts both signed and unsigned 64-bit integers to 32-bit integers: https://godbolt.org/z/7orpn5

The construction of a `std::uint64_t` using braces tells the compiler that narrowing is prohibited.  So this is one of the few ways we can have confidence that no narrowing will occur.

scottschurr commented on src/ripple/app/misc/impl/TxQ.cpp:
```diff
@@ -353,75 +374,62 @@ TxQ::isFull() const
+    // transaction goes in front of any queued transactions.  Enables
+    // recovery of open ledger transactions and stuck transactions.
+    if (tx.getSeqOrTicket() >= accountIter->second.transactions.rbegin()->first)
+        return telCAN_NOT_QUEUE_FULL;
```
Thanks for donating the solution.  I'm going with the definitely safe approach.

scottschurr replied:
The good news is that this change caused a unit test failure.  The (first) location that failed is here:

https://github.com/ripple/rippled/pull/3271/files#diff-4cd96b2a7559d60e04070827ccede718R544

That line now returns `telCAN_NOT_QUEUE_FULL` as does the following line.  I think the failure is correct under the new rules, which means we're better at identifying an account's queue being full.  I've re-tuned the unit test to match the change.

If you think the new failure is in error please let me know.  Thanks.

ximinez replied:
The link isn't taking me to the right spot, but I did encounter and flag at least one test that was allowing too many transactions into the queue, so I'm glad I was right. :)

scottschurr replied:
Sorry about the link.  Yeah, it was just one test that had 11 failures after that change.  After the change there were two additional tests that were prevented from entering the queue.  So I think the change is good.

scottschurr commented on src/ripple/app/misc/impl/TxQ.cpp:
```diff
@@ -745,189 +891,226 @@ TxQ::apply(Application& app, OpenView& view,
-        {
-            auto& txQAcct = accountIter->second;
-            auto const aSeq = (*sle)[sfSequence];
+        if (txSeqOrT.value() < acctSeqOrT.value())
```
I believe that `<` is correct.  The current account sequence can be turned into a `TicketSequence` if a `TicketCreate` transaction uses a `Ticket`.  So I think the correct response is `terPRE_TICKET` which is what you'll get from this code.

ximinez replied:
Ah, yes, good point. Withdrawn. :)

ximinez commented on src/test/app/Ticket_test.cpp:
```diff
@@ -25,369 +25,376 @@ namespace ripple {
+        }
+        {
+            // Another way to exceed the threshold where tickets can no longer
+            // beadded to an account.
```
This comment doesn't seem to exist anymore, so nevermind.

scottschurr commented on src/ripple/app/misc/impl/TxQ.cpp:
```diff
@@ -745,189 +891,226 @@ TxQ::apply(Application& app, OpenView& view,
+                TER const ter {
+                    prevIter->second.consequences().couldBeNext (txSeqOrT)};
+                if (!isTesSuccess (ter))
+                    return {ter, false};
```
Nice!  This change helped me find a bug in `nextQueuableSeq()`.  As is noted elsewhere in `apply` sometimes transactions are present both in the queue and in the ledger.  When that happens `nextQueuableSeq()` would do the wrong thing.  So I've recoded `nextQueuableSeq()` like this:
```
// The passed in proposedSeq is typically (but not required to be) the
// sequence of the account root of account.  The goal is to return a
// SeqProxy for a sequence that will fill the next available hole
// in the queue for the passed in account.
//
// Typically there will be no queued transactions.  In that case the
// proposedSeq is returned as a SeqProxy.
//
// If there are queued transactions for the account then the first viable
// sequence number, that is not used by a transaction in the queue, must
// be found and returned.
SeqProxy
TxQ::nextQueuableSeqImpl (
    ReadView const& view,
    AccountID const& account,
    std::uint32_t proposedSeq,
    std::lock_guard<std::mutex> const&) const
{
    SeqProxy const proposed = SeqProxy::sequence (proposedSeq);

    // If the account is not in the queue, the proposed start is good enough.
    auto const accountIter = byAccount_.find(account);
    if (accountIter == byAccount_.end() ||
            accountIter->second.transactions.empty())
        return proposed;

    // If the account is not in the ledger the proposed start is good enough.
    auto const sleAccount = view.read (keylet::account (account));
    if (! sleAccount)
        return proposed;

    TxQAccount::TxMap const& acctTxs = accountIter->second.transactions;

    // Ignore any sequence-based queued transactions that slipped into the
    // ledger while we were not watching.  This does actually happen in the
    // wild, but it's uncommon.
    SeqProxy const acctSeqProx =
        SeqProxy::sequence ((*sleAccount)[sfSequence]);

    TxQAccount::TxMap::const_iterator txIter =
        acctTxs.lower_bound (acctSeqProx);

    if (txIter == acctTxs.end() ||
            !txIter->first.isSeq() ||
            txIter->first != acctSeqProx)
        // Either there are no queued sequence-based transactions equal to or
        // following acctSeqProx or acctSeqProx is not currently in the queue.
        // So acctSeqProx is as good as it gets.
        return acctSeqProx;

    // There are sequence-based transactions queued that follow acctSeqProx.
    // Locate the first opening to put a transaction into.
    SeqProxy attempt = txIter->second.consequences().followingSeq();
    while (++txIter != acctTxs.cend())
    {
        if (attempt < txIter->first)
            break;

        attempt = txIter->second.consequences().followingSeq();
    }
    return attempt;
}
```
Let me know if you spot any problems in that.  Thanks!

ximinez replied:
The only thing that jumps out at me is the same idea I put out in another comment instead of looking it up in the middle of the function, pass in `SLE const& sleAccount`, and get the information `AccountID` and sequence out of it. I know the comment says the proposed seq is not required to be the account sequence, but why allow that, when all it does is open up more potential to give the wrong answer? I'd rather know that the function is more likely to give me what it promises to give me: the next queuable sequence.

scottschurr commented on src/ripple/app/misc/impl/TxQ.cpp:
```diff
@@ -1369,19 +1659,19 @@ TxQ::getTxRequiredFeeAndSeq(OpenView const& view,
+            if (seqOrTicket.value() >= availableSeq)
+                availableSeq = seqOrTicket.value() + 1;
         }
     }
```
Good catch!  Yes, I replaced the loop with a call to `nextQueuableSeqImpl()`.

scottschurr commented on src/ripple/app/misc/impl/TxQ.cpp:
```diff
@@ -745,189 +891,226 @@ TxQ::apply(Application& app, OpenView& view,
+                return {ter, false};
+
+            requiresMultiTxn = true;
+        }
```
If this is rewritten as a lambda then the `return {ter, false}` gets more complicated to handle.  I think it's probably better the way it is.

ximinez replied:
Yeah, good point.

scottschurr commented on src/ripple/app/misc/impl/TxQ.cpp:
```diff
@@ -1242,8 +1472,6 @@ TxQ::accept(Application& app,
         if (feeLevelPaid >= requiredFeeLevel)
         {
-            auto firstTxn = candidateIter->txn;
-
```
With the introduction of tickets "remove the last item of account" makes less sense than it used to.  But let's try to understand the motivation for the `dropPenalty`.

- If the idea is to simply open up another spot in the queue, then dropping an arbitrary ticket-transaction (actually the "last" one) is as good as dropping the last sequence-transaction.

- If the idea is to provide a punishment for a failing transaction in a nearly full queue then, again, dropping an arbitrary ticket-transaction (actually the "last" one) is as good as dropping the last sequence-transaction.

In either of those cases it seems like the _best_ option would be to drop the transaction with the lowest fee.  But we're in a rare circumstance, so I don't think it's worth the effort to search for the prefect transaction to remove.

Therefore I think, as long as we want to drop one of the account's queued transactions, dropping either the "last" ticket or the last sequence serves the purpose.  Therefore I'm not planning to change that part of the logic.  Please let me know if you disagree.  Thanks.

While we're looking at that block, I'm nervous about the
```
assert(endIter != candidateIter);
```
With tickets I'm no longer confident that will necessarily be the case.  My preference would be to replace the `assert` with
```
if (endIter != candidateIter)
    erase(endIter);
```
I feel like that's a bit more future proof.  I'll go ahead and make that change.  But please let me know if you disagree.  Thanks again.

ximinez replied:
Great points. The idea is definitely to open up a spot in the queue for by dropping a less expensive transaction to allow a more expensive one in. Now that you mention it, I think the _bestest_ option would be to drop the transaction with the lowest fee from the set of { ticket-transactions, the last sequence-transaction }. I wouldn't want to drop a ticket from the middle of the sequences, because that would kill all the others after it. For tickets, that's not true.

However, all that said, simplicity has value, and given that this is a very rare circumstance, I am fine with leaving the drop as-is.

I definitely agree about the possibility that `endIter == candidateIter`, since Tickets are applied in fee order, and the last ticket by sequence order could easily be the one that pays most for this account. Your change looks good and safe.

scottschurr commented on src/ripple/rpc/handlers/AccountInfo.cpp:
```diff
@@ -126,19 +126,43 @@ Json::Value doAccountInfo (RPC::JsonContext& context)
 
+                // We expect txs to be returned sorted by SeqOrTicket.  Verify
+                // that with a couple of asserts.
+                std::uint32_t lastSeqValue = 0;
+                std::uint32_t lastTicketValue = 0;
```
:+1: Good call.  Thanks.

scottschurr commented on src/ripple/app/misc/impl/TxQ.cpp:
```diff
@@ -1321,6 +1549,68 @@ TxQ::accept(Application& app,
+// be found and returned.
+SeqOrTicket
+TxQ::nextQueuableSeq (
+    ReadView const& view,
```
:+1:  And I went will passing only the `std::shared_ptr<SLE const> const&`

scottschurr commented on src/test/app/TxQ_test.cpp:
```diff
@@ -354,6 +374,244 @@ class TxQ_test : public beast::unit_test::suite
+        // Alice requests tickets, but that transaction is queued.  So
+        // Alice can't queue ticketed transactions yet.
+        std::uint32_t const tkt1 {env.seq(alice) + 1};
+        env(ticket::create (alice, 250), seq(tkt1 - 1), queued);
```
Agreed.  But, for me, it is easier to see what the test is doing if the sequence is explicitly specified.  Think of it as a comment that is actually executed :wink:

scottschurr commented on src/test/app/TxQ_test.cpp:
```diff
@@ -610,24 +869,15 @@ class TxQ_test : public beast::unit_test::suite
-
-        // Carol fills the queue, but can't kick out any
-        // transactions.
+            seq(seqBob), ter(telCAN_NOT_QUEUE_BLOCKS));
```
So I'm digging into this test now.  I don't quite have the changes you want yet.  Here's what I did:

1. Rather than filling bob's queue prior to putting in the `regkey`, I'm filling Alice's queue.
2. Then I submit bob's `regkey`, expecting it to go into the queue.  Instead it goes straight into the ledger.  The new `tryDirectSubmit` sends the `regkey` straight in because it has a very high (but not infinite) `feeLevelPaid`.
3. I do more stuff with carol, but that's irrelevant since the `regkey` is not in the queue.

Since `regkey` is a blocker and must be alone in bob's queue I'm unsure how to keep bob's `regkey` in the queue.  Maybe this indicates a problem with the `tryDirectSubmit` implementation?

Thanks for your thoughts.

ximinez replied:
@scottschurr  We've discussed this offline. Here's a very short summary of the conversation.

Basically, there are feasible ways to solve this problem, specifically in the test.
1. Override the value of `zero_basefee_transaction_feelevel` for this test to be something more in the range of Alice's queued transactions, causing the tx to go into the queue.
2. Refactor the way these zero-fee transactions are handled internally so that the system never has to deal with the divide-by-zero issue. One way to implement this would be for `CalculateBaseFee` to return the actual base fee, and an "effective fee paid". In most cases, the base fee would remain unchanged, and the "effective fee" would be the actual fee. In this special case, setting a regkey with no fee required, the base fee would be the default fee as if this was a normal transaction, but the effective fee would be `the actual fee + toDrops(base fee)`. Any place that uses the base fee and the transaction fee would instead use the effective fee. Everywhere else, particularly when the fee is actually charged, would use the actual fee.

scottschurr commented on src/ripple/app/tx/impl/Transactor.cpp:
```diff
@@ -312,49 +312,58 @@ TER Transactor::consumeSeqOrTicket (SLE::pointer const& sleAccount)
-        }
+    // Update the account root's TicketCount.  If the ticket count drops to
+    // zero remove the (optional) field.
+    auto sleAccount = view.peek (keylet::account(account));
```
I explored this.  The complication is that `Transactor::ticketDelete()` is also being called from DeleteAccount.cpp.  I put some effort into keeping the number of parameters that DeleteAccount needs to pass minimized, since it also impacts three additional related functions.  So I could make the change you're suggesting.  But in deference to DeleteAccount I'd prefer not too.

Yeah, I acknowledge that `Keylet` computation is expensive.  So if you would like we can revisit this discussion.  I'm to dead set that I'm making the right call.

ximinez replied:
Actually, I realized that the way you're passing in the `ticketIndex`, you've already done the hard work, so creating the `Keylet` is computationally cheap. So this is fine as-is.

scottschurr commented on src/ripple/app/tx/applySteps.h:
```diff
@@ -38,6 +38,138 @@ isTecClaimHardFail(TER ter, ApplyFlags flags)
+    }
+
+    // Could nextSeqOrT immediately follow this?
+    NotTEC couldBeNext (SeqOrTicket nextSeqOrT) const
```
Huh.  I discovered that this method is no longer being called by anyone.  So I removed it.

ximinez replied:
LOL. I remember seeing it at one point, so it was useful once, but probably got replaced by something more useful. I bet it was `followingSeq`.

scottschurr commented on src/ripple/app/tx/applySteps.h:
```diff
@@ -38,6 +38,138 @@ isTecClaimHardFail(TER ter, ApplyFlags flags)
+    SeqOrTicket followingSeq () const
+    {
+        return SeqOrTicket {
+            seqOrT_.isSeq() ? SeqOrTicket::seq : SeqOrTicket::ticket,
```
I looked at the options here.  I ended up going with adding a named `SeqProxy::advanceBy()` method.  I didn't want to introduce an `operator+` which would suggest this is a common operation.  It's not.  It's really pretty unusual.  I only found three places that can use `advanceBy()`: one here and two places in `TxQ_test::testClearQueuedAccountTxs()`.  There may be others...

scottschurr commented on src/ripple/protocol/impl/Feature.cpp:
```diff
@@ -93,7 +93,6 @@ detail::supportedAmendments ()
     static std::vector<std::string> const supported
     {
         "MultiSign",                   // Unconditionally supported.
-        "Tickets",       // Looks unused, but do not remove; Servers will be amendment blocked.
```
I've learned more about amendments.  Servers only become amendment blocked if the removed amendment was enabled.  If the amendment was never enabled (as is the case with Tickets) then removing the amendment will not block anybody.

scottschurr commented on src/ripple/app/tx/impl/Escrow.cpp:
```diff
@@ -90,39 +90,42 @@ static inline bool after (NetClock::time_point now, std::uint32_t mark)
-    return tx[sfAmount].xrp();
+    auto const& amount {tx[sfAmount]};
+    return amount.native() && !amount.negative() ?
+        amount.xrp() : beast::zero;
```
Somewhat surprisingly, it turns out that the check is necessary and it is not okay to add an `assert` here.  The issue is that `Consequences` are returned from `preflight()` unconditionally.  So no matter what error code `preflight()` returns, and even if `preflight()` detects a bad `amount`, `makeTxConsequences()`will be called.  Therefore `makeTxConsequences()` must be prepared to handle bad values.

scottschurr replied:
It could be done, but I'm not convinced the effort and complexity would pay off.  Here's where `TxConsequences` are being collected now: https://github.com/ripple/rippled/blob/40c9999d18e66438e47ff7d76959b58d50ba2c3d/src/ripple/app/tx/impl/applySteps.cpp#L48-L73

So the `TxConsequences` are computed unconditionally.  At the moment we don't even know which is called first: `T::preflight` or the `TxConsequences` constructor.

To change this we'd change the three `invoke_preflight_helper` functions to call `preflight` first and, based on the return code, call one of two `TxConsequences` constructors.  That's certainly doable.

What I'm uncertain of is the benefit.  To keep `TxConsequences` low cost we'd like them to have half-way reasonable responses to all the questions they can answer.  The easiest way to get that answer is to consult the actual transaction in question.

Maybe I can turn the question around.  What would you expect a `TxConsequences` to look like when `preflight` returns a `tel` code?  Is that something that can be generalized?  Perhaps we should be returning `optional<TxConsequences>`?

ximinez commented on src/test/app/TxQ_test.cpp:
```diff
@@ -2555,6 +2554,127 @@ class TxQ_test : public beast::unit_test::suite
+
+        // Finally we can fill in the entire gap.
+        env(noop(alice), seq(aliceSeq + 18), ter(terQUEUED));
+        checkMetrics(env, 14, 40, 5, 4, 256);
```
Add one more `env(noop(alice), seq(aliceSeq + 20), ter(telCAN_NOT_QUEUE_FULL));` after you've filled the gap to verify that you can't queue anything at the end now that the gap is full.

ximinez commented on src/ripple/app/tx/impl/Transactor.cpp:
```diff
@@ -90,6 +90,17 @@ preflight1 (PreflightContext const& ctx)
         return temBAD_SIGNATURE;
     }
 
+    // An AccountTxnID field constrains transactions more than the Sequence
+    // field.  Tickets, on the other hand, reduce ordering constraints.
+    // A transaction that includes both fields is difficult to get into the
+    // ledger and makes it hard for the TxQ to know the likelihood of a
+    // transactions success.
```
I suggest a rephrasing:
```
    // An AccountTxnID field constrains transactions more than the Sequence
    // field.  Tickets, on the other hand, reduce ordering constraints.
    // Because one counteracts the other, the combination is unsupported,
    // and treated a malformed.
```

scottschurr commented:
At this point I think I've addressed all of @ximinez's comments.  If you spot any that I missed please let me know.

I'm gettin' close, but not done yet.  Running code coverage locally showed holes in TxQ coverage in two corner cases that are affected by Tickets.  I'm gonna see if I can write tests that exercise those two corner cases.

scottschurr commented on src/ripple/protocol/impl/Feature.cpp:
```diff
@@ -131,7 +130,7 @@ detail::supportedAmendments ()
         "RequireFullyCanonicalSig",
         "fix1781",
-        "TicketBatch",
+//        "TicketBatch",    // Commented out to prevent automatic enablement
```
This, or something like it, is a good idea.  But I'd prefer to see it done in a separate pull request.

ximinez commented on src/ripple/app/misc/impl/TxQ.cpp:
```diff
@@ -1337,9 +1521,8 @@ TxQ::accept(Application& app,
                         ". Removing last item of account " <<
                         account.account;
                     auto endIter = byFee_.iterator_to(dropRIter->second);
-                    assert(endIter != candidateIter);
-                    erase(endIter);
-
+                    if (endIter != candidateIter)
+                        erase(endIter);
```
I just looked at this block again, and I'm thinking about the reasons why dropping the last transaction was so important: When all you have is sequence transactions, you can't drop the current one, because you're guaranteed to wedge the later ones. If you drop the last sequence transaction, there's a chance, however small, for the others to succeed. Now with tickets, if the current transaction is ticket based, there's no reason not to drop the current ticket based transaction - it won't affect the ones "after" it.

So here's my suggestion. I'm not sure if it breaks any tests.

```
                if (account.dropPenalty && account.transactions.size() > 1 &&
                    isFull<95>())
                {
                    /* The queue is close to full, this account has multiple
                        txs queued, and this account has had a transaction
                        fail.
                    */
                    if (candidateIter->seqProxy.isTicket())
                        // Since this is a ticket, order does matter.
                        // Drop this one
                        candidateIter = eraseAndAdvance(candidateIter);
                    else
                    {
                        /* Even though we're giving this transaction another
                           chance, chances are it won't recover. So we don't
                           make things worse, drop the _last_ transaction for
                           this account.
                        */
                        auto dropRIter = account.transactions.rbegin();
                        assert(dropRIter->second.account == candidateIter->account);
                        JLOG(j_.warn()) << "Queue is nearly full, and transaction "
                                        << candidateIter->txID << " failed with "
                                        << transToken(txnResult)
                                        << ". Removing last item of account "
                                        << account.account;
                        auto endIter = byFee_.iterator_to(dropRIter->second);
                        if (endIter != candidateIter)
                            erase(endIter);
                        ++candidateIter;
                    }
                }
                else
                    ++candidateIter;
```

scottschurr commented:
I think this most recent commit adds the necessary TxQ unit tests and addresses the remaining concerns identified so far.

ximinez commented on src/test/app/TxQ_test.cpp:
```diff
@@ -3967,6 +3967,293 @@ class TxQ_test : public beast::unit_test::suite
+        checkMetrics(env, 60, 185, 38, 37, 256);
+
+        // The next close should cause the in-ledger amendments to change.
+        // Alice's queued transactions have a cashed PreflightResult
```
Nit: "cached"

ximinez commented:
:tada: :tada: :tada:

seelabs commented on src/ripple/proto/org/xrpl/rpc/v1/transaction.proto:
```diff
@@ -55,8 +55,9 @@ message Transaction
         SignerListSet signer_list_set = 28;
 
-        TrustSet trust_set = 29;
+        TicketCreate ticket_create = 30;
```
I know this was done intentionally, but I vote to move this to the end. I'd rather list them in id order and break alphabetical order rather than the other way around. (No change needed, just stating a preference).

scottschurr replied:
Thanks @seelabs.  @cjcobb23, do you have a preference for this ordering?  Since these proto definitions were primarily your work I'd like to do it the way you prefer.

cjcobb23 replied:
I prefer alphabetical order here. It's just easier to read. The ids are only really used internally and don't really mean anything to humans.

scottschurr commented:
@cjcobb23 and @HowardHinnant, would you like an opportunity to review the most recent changes before I squash this all down and rebase it?  I'll wait until I've heard from you.  Thanks.

HowardHinnant commented:
Looks good to me. 👍

cjcobb23 commented:
@scottschurr you can squash

scottschurr commented:
Squashed and rebased.
